Add authentication with credentials file

stjepang opened the pull request:
This PR adds `ConnectionOptions::with_credentials()` that allows you to specify a `.cred` file to authenticate.

Connection options handling is becoming unwieldy - would you be okay with me doing a little bit refactoring around that? In particular, I'd like to move everything options-related into a dedicated file.

stjepang commented on src/shared_state.rs:
```diff
@@ -370,3 +416,11 @@ fn empty_or_none(field: &Option<&String>) -> bool {
 }
+
+#[inline]
+fn empty_or_none_owned(field: &Option<String>) -> bool {
```
Would be better if we didn't need this function. The fact that `Connect<'_>` has a lifetime and is borrowing temporary variables makes the code a bit more convoluted that it needs to be, I think.

What if instead of having fields of type `Option<&String>` we simply had `Option<String>` (talking about `user`, `pass`, `auth_token`, etc)? It's not like saving a few allocations in the options struct saves us any tangible performance...

derekcollison replied:
Makes sense.

derekcollison commented on src/shared_state.rs:
```diff
@@ -111,13 +123,30 @@ impl Server {
+                    .map_err(|err| Error::new(ErrorKind::InvalidData, err))?;
+                nkey = Some(key_pair);
+
+                jwt = Some(user_jwt);
```
Where do we sign the nonce?

derekcollison replied:
What we want ideally is that this is loaded on demand only when needed to form CONNECT {}.

derekcollison commented on src/lib.rs:
```diff
@@ -557,9 +586,16 @@ pub struct Connection {
     UserPass(String, String),
+
+    /// Authenticate using a `.creds` file.
+    Credentials(PathBuf),
```
This in the Go client is simply and concatenation of JWT and pl(nkey) which we will want to also support as separate ones. Look at Go client for pointers.

stjepang replied:
Done - this variant now looks like this:
```rust
/// Authenticate using a `.creds` file.
Credentials {
    /// Securely loads the user JWT.
    jwt_cb: Arc<dyn Fn() -> io::Result<String> + Send + Sync>,
    /// Securely loads the nkey and signs the nonce passed as an argument.
    sig_cb: Arc<dyn Fn(&[u8]) -> io::Result<SecureString> + Send + Sync>,
},
```

derekcollison replied:
Looks good!

derekcollison commented on src/shared_state.rs:
```diff
@@ -161,11 +190,24 @@ impl Server {
+            let nkey = nkey.unwrap();
+            let sig = nkey.sign(info.nonce.as_bytes()).unwrap();
+            let sig = base64_url::encode(&sig);
+            connect_op.sig = Some(sig);
```
Try to wipe memory as best as we can once the CONNECT is sent.

spacejam replied:
Implemented buffer & string zeroing via the new `SecureVec` & `SecureString` structs that scramble their data on Drop

derekcollison commented on src/lib.rs:
```diff
@@ -349,6 +349,35 @@ impl ConnectionOptions<options_typestate::NoAuth> {
+    /// # fn main() -> std::io::Result<()> {
+    /// let nc = nats::ConnectionOptions::new()
+    ///     .with_credentials("path/to/my.creds")
+    ///     .connect("demo.nats.io")?;
```
Make `connect.ngs.global`

stjepang commented:
Okay, this is ready for review now. Changes:

- We added secure wiping of sensitive data.
- Refactored the code a bit to untangle the connection process and added more comments.
- Modeled `Credentials` after the Go client - now we have two callbacks in it - one to get the user JWT and one to sign a nonce.

derekcollison commented:
Looks good.

spacejam commented on src/secure_wipe.rs:
```diff
@@ -0,0 +1,104 @@
+        }
+        compiler_fence(SeqCst);
+    }
+}
```
this is broken without the push function that I included originally

spacejam replied:
when parser::expect_info reads in the ServerInfo, it may realloc, causing non-scrambled copies of memory to float around in the vacated allocations

spacejam commented on src/connect.rs:
```diff
@@ -0,0 +1,207 @@
+
+    /// Connection username (if `auth_required` is set)
+    #[serde(skip_serializing_if = "is_empty_or_none")]
+    pub user: Option<String>,
```
`SecureString`

spacejam commented on src/connect.rs:
```diff
@@ -0,0 +1,207 @@
+
+    /// User's JWT.
+    #[serde(rename = "jwt", skip_serializing_if = "is_empty_or_none")]
+    pub user_jwt: Option<String>,
```
SecureString

spacejam commented on src/connect.rs:
```diff
@@ -0,0 +1,207 @@
+
+    /// Signed nonce, encoded to Base64URL.
+    #[serde(rename = "sig", skip_serializing_if = "secure_is_empty_or_none")]
+    pub signature: Option<SecureString>,
```
SecureString

spacejam commented on src/connect.rs:
```diff
@@ -0,0 +1,207 @@
+    // Fill in the info that authenticates the client.
+    match &options.auth {
+        AuthStyle::UserPass(user, pass) => {
+            connect_info.user = Some(user.to_string());
```
SecureString
